plugin/audit_log/filter.c

```diff
@@ -0,0 +1,242 @@
+*/
+static
+void account_init(account *acc, const char *user, size_t user_length,
+                                const char *host, size_t host_length)
+{
+  DBUG_ASSERT(user_length + host_length + 2 <= sizeof(acc->name));
+
+  memcpy(acc->name, user, user_length);
+  memcpy(acc->name + user_length + 1, host, host_length);
+  acc->name[user_length]= '@';
+  acc->name[user_length + host_length + 1]= 0;
+  acc->length= user_length + host_length + 1;
+}
+
+/*
+  Allocate memory and initialize new account
+*/
+
+static
+account *account_create(const char *user, size_t user_length,
+                        const char *host, size_t host_length)
+{
+  account *acc= (account *) my_malloc(sizeof(account),
+                                      MYF(MY_FAE | MY_ZEROFILL));
+
+  account_init(acc, user, user_length, host, host_length);
+
+  DBUG_ASSERT(user_length + host_length + 2 <= sizeof(acc->name));
+
+  memcpy(acc->name, user, user_length);
+  memcpy(acc->name + user_length + 1, host, host_length);
+  acc->name[user_length]= '@';
+  acc->name[user_length + host_length + 1]= 0;
+
+  return acc;
+}
+
+/*
+  Get account key
+*/
+uchar* account_get_key(const account *acc, size_t *length,
+                       my_bool not_used __attribute__((unused)))
+{
+  *length= acc->length;
+  return (uchar*) acc->name;
+}
+
+/*
+  Remove enclosing quotes from string if any.
+*/
+static
+void unqote_string(char *string, size_t *string_length)
+{
+  if (string[0] == '\'' && string[*string_length - 1] == '\'')
+  {
+    memmove(string, string + 1, *string_length);
+    *string_length= *string_length - 2;
+    string[*string_length]= 0;
+  }
+}
+
+/*
+  Parse comma-separated list of accounts and add it into account list.
+  Empty user name is allowed.
+*/
+static
+void account_list_from_string(HASH *hash, const char *string)
+{
+  size_t entry_length;
+  char *string_copy= my_strdup(string, MYF(MY_FAE));
+  char *entry= string_copy;
+  int string_length= strlen(string_copy);
+  my_bool quote;
+  char user[USERNAME_LENGTH], host[HOSTNAME_LENGTH];
+  size_t user_length, host_length;
+
+  my_hash_reset(hash);
+
+  while (entry - string_copy < string_length)
+  {
+    quote= FALSE;
+
+    while (*entry == ' ')
+      entry++;
+
+    entry_length= 0;
+    while (((entry[entry_length] != ' ' && entry[entry_length] != ',') || quote)
+           && entry[entry_length] != 0)
+    {
+      if (entry[entry_length] == '\'')
+        quote= !quote;
+      entry_length++;
+    }
+
+    entry[entry_length]= 0;
+
+    parse_user(entry, entry_length, user, &user_length, host, &host_length);
+    unqote_string(user, &user_length);
+    unqote_string(host, &host_length);
+
+    my_hash_insert(hash,
+        (uchar*) account_create(user, user_length, host, host_length));
+
+    entry= entry + entry_length + 1;
+  }
+
+  my_free(string_copy);
+}
+
+/* public interface */
+
+void audit_log_filter_init()
+{
+#ifdef HAVE_PSI_INTERFACE
+  mysql_rwlock_register("server_audit", all_rwlock_list,
```

Comment:
Do you mean category? There is a problem here already.
```c++
#ifdef HAVE_PSI_INTERFACE
  if(PSI_server)
  {
    PSI_server->register_mutex("server_audit",
                               mutex_key_list, array_elements(mutex_key_list));
    PSI_server->register_cond("server_audit",
                              cond_key_list, array_elements(cond_key_list));
  }
#endif /* HAVE_PSI_INTERFACE */
```
and in `logger.h` (weird, no?):
```
#define PSI_CATEGORY "audit_logger"
```
and then in `audit_log.c`
```
  mysql_memory_register(PSI_CATEGORY, all_audit_log_memory, count);
```
But thats in 5.7.

In 5.6 we have `"server_audit"` in `buffer.c` and `"audit_logger"` in `file_logger.c` (which is taken from MariaDB).

I can address it in this PR and make sure it consistent in 5.7 as well.